### ISSUE TITLE
Non-trivial changes to Flux.1 and Fibo required to achieve CodeRabbit approval for test-tiny

### ISSUE THREAD

**Issue by ianscrivener** [filipstrand/mflux#617]
Investigating doing tiny tests with Claude Code showed that some more significant changes to Flux.1 and Fibo were needed to make CodeRabbit happy. These changes will require doing full testing with Flux.1 and Fibo.

This feels like it's above my pay grade and would prefer that @plz12345 or @fxd0h take a look at this please.

<sub>Claude wrote the following PRD</sub>

--- 

# PRD: Parameterize `AdaLayerNormZeroSingle` / `TimestepEmbedder` (shared FLUX.1 + FIBO), then add tiny tests for both

## Background

`src/mflux/models/flux/model/flux_transformer/ada_layer_norm_zero_single.py` and
`src/mflux/models/flux/model/flux_transformer/timestep_embedder.py` are shared
between **FLUX.1** and **FIBO** (FIBO imports these FLUX.1 classes directly rather
than owning its own copies — confirmed via `grep -rln "AdaLayerNormZeroSingle\|TimestepEmbedder" src/mflux/`).

Both classes hardcode `dim=3072` with **zero constructor parameters**:

```python
# ada_layer_norm_zero_single.py
class AdaLayerNormZeroSingle(nn.Module):
    def __init__(self):
        super().__init__()
        self.linear = nn.Linear(3072, 3 * 3072)
        self.norm = nn.LayerNorm(dims=3072, eps=1e-6, affine=False)
```

```python
# timestep_embedder.py
class TimestepEmbedder(nn.Module):
    def __init__(self):
        super().__init__()
        self.linear_1 = nn.Linear(256, 3072)
        self.linear_2 = nn.Linear(3072, 3072)
```

This is why:
- The **FLUX.1 tiny model-saving test could not be written at all** (see
  `.claude/notes/tiny-tests.md` — FLUX.1's `T5Encoder`, `CLIPEncoder`, and
  `Transformer` all hardcode dims with no shrink knobs; this pair is part of that
  same problem, inside `Transformer`'s single-block path).
- The **FIBO tiny test** (`tests/model_saving/test_tiny_model_saving_fibo.py`,
  currently held back off `tiny-tests-2`) has to define a test-local
  `_TinyFiboTransformer` that composes FIBO's real leaf classes at toy dims — but
  these two leaves can't shrink, so they're instantiated at full production size
  (3072) inside an otherwise-tiny (dim=128) fixture. CodeRabbit flagged this:

  > `FiboSingleTransformerBlock(..., dim=128)` still creates
  > `AdaLayerNormZeroSingle()` at width 3072. `BriaFiboTimestepProjEmbeddings()`
  > also creates a 3072-wide `TimestepEmbedder`. The two single blocks and time
  > embedding add 66,871,296 parameters (~255 MiB at fp32) before the remaining
  > test components. The helper creates this fixture twice and randomizes every
  > parameter, making a "fast" test unnecessarily expensive.

Note both call sites already thread a `dim` param through everything else — they
just don't pass it to these two:

```python
# fibo/model/fibo_transformer/single_transformer_block.py
def __init__(self, layer, dim: int = 3072, num_attention_heads=24, attention_head_dim=128, mlp_ratio=4.0):
    ...
    self.norm = AdaLayerNormZeroSingle()  # <- dim not passed through
```

```python
# fibo/model/fibo_transformer/time_embed.py
def __init__(self, embedding_dim: int = 3072, time_theta: int = 10000):
    ...
    self.timestep_embedder = TimestepEmbedder()  # <- embedding_dim not passed through
```

```python
# flux/model/flux_transformer/single_transformer_block.py and time_text_embed.py
self.norm = AdaLayerNormZeroSingle()          # same issue, FLUX.1 side
self.timestep_embedder = TimestepEmbedder()   # same issue, FLUX.1 side
```

## Why this is its own workstream, not a quick fix

Mechanically this is a small, reviewable diff (add a `dim: int = 3072` parameter,
replace the hardcoded `3072` literals with it, thread it through the two callers
in each of `flux/` and `fibo/`). But `mflux` is an explicitly **line-by-line
numerical port** from PyTorch/diffusers — any change to a class in FLUX.1's and
FIBO's actual forward pass needs to be proven bit-identical at the default
(3072) width, not just "reads like a no-op refactor." That verification pass —
export-then-compare against the PyTorch reference per the `mflux-debugging`
skill — is the bulk of the real work here, not the code change itself.

## Goal

1. Parameterize `AdaLayerNormZeroSingle` and `TimestepEmbedder` with a `dim: int
   = 3072` constructor argument, defaulting to current production behavior.
2. Thread `dim` through both call sites in `flux/model/flux_transformer/` and
   both call sites in `fibo/model/fibo_transformer/`.
3. Prove numerical parity at the default width for **both** FLUX.1 and FIBO
   (not just "tests still pass" — actual tensor-level comparison against the
   PyTorch/diffusers reference, per `mflux-debugging`).
4. Once (1)–(3) land, write/restore the tiny model-saving tests:
   - **FLUX.1**: currently blocked entirely (no tiny test exists). This fix
     removes one blocker but does *not* fully unblock it — `T5Encoder` and
     `CLIPEncoder` are separately unparameterized and out of scope here. Re-run
     the `mflux-model-tiny-test` skill's blocker analysis after this lands to
     confirm whether FLUX.1 is unblocked or just less-blocked.
   - **FIBO**: re-add `tests/model_saving/test_tiny_model_saving_fibo.py`
     (currently held back on branch `tiny-tests-2`, commit removing it exists —
     recover the file from git history), update `_TinyFiboTransformer` to pass
     `dim=128` through to the now-parameterized `AdaLayerNormZeroSingle` and
     `TimestepEmbedder`, and confirm CodeRabbit's flagged param count drops
     accordingly.

## Non-goals

- Not fixing FLUX.1's `T5Encoder`/`CLIPEncoder` (separate, larger effort — no
  constructor params at all, ~4.7B and ~123M params respectively at their one
  fixed size). Tracked separately in `.claude/notes/tiny-tests.md`.
- Not touching Z-Image's or ERNIE's timestep embedders — they're already
  independently parameterized, separate classes under their own model
  directories, unaffected by this change.

## Verification plan (per `mflux-debugging` skill)

Do this for **both** FLUX.1 and FIBO, since both consume the shared classes:

1. Pick one deterministic repro per model (fixed seed, prompt, height/width,
   steps, guidance).
2. Export reference tensors from the PyTorch/diffusers side at the checkpoints
   that touch these two classes:
   - Timestep projection output → `TimestepEmbedder` output (per model)
   - `AdaLayerNormZeroSingle` output (`hidden_states`, `gate`) for at least one
     single-transformer-block, ideally the first and last
3. Run the same repro in mflux **after** the parameterization change (still at
   `dim=3072` default) and export the matching tensors.
4. Compare with `mflux-debugging`'s standard checkpoint-comparison approach:
   shape+dtype match, `max_abs_diff`/`mean_abs_diff`/`max_rel_diff` against
   `atol=1e-5, rtol=1e-5` (fp32) or `atol=1e-2, rtol=1e-2` (bf16/fp16) as
   appropriate, inspecting actual first-N values rather than trusting summary
   stats alone.
5. Also do one full end-to-end image generation on each model, before/after,
   same seed — confirm output images are unchanged (or diff imperceptibly
   within known fp precision noise).
6. Do **not** rely on integer-seed matching for RNG — inject the exact initial
   latents per the skill's RNG warning if any step requires seed-level parity.

Only after (3)+(5) pass for both models should the tiny tests in Goal #4 be
written/restored.

## Acceptance criteria

- [ ] `AdaLayerNormZeroSingle(dim: int = 3072)` and `TimestepEmbedder(dim: int =
      3072)` (or equivalently named param) merged, with both FLUX.1 and FIBO
      call sites passing their existing `dim`/`embedding_dim` through.
- [ ] Tensor-level parity confirmed and documented (paste the comparison output
      or a summary) for both FLUX.1 and FIBO at default width.
- [ ] End-to-end image parity confirmed for both models at a fixed seed/prompt.
- [ ] `just test-fast` / `just test` green, no regressions.
- [ ] FIBO tiny test restored and green, with the previously-flagged 3072-width
      components now actually shrunk to the fixture's toy dimensions (verify
      CodeRabbit's specific concern is resolved — recompute param count).
- [ ] FLUX.1 tiny test: re-assess blocker status; write it if now unblocked,
      otherwise update `.claude/notes/tiny-tests.md` with the narrowed
      remaining blocker (`T5Encoder`/`CLIPEncoder`).

## References

- `.claude/notes/tiny-tests.md` — coverage tracker, FLUX.1 blocker writeup,
  FIBO workaround writeup, this issue's origin.
- `.cursor/skills/mflux-model-tiny-test/SKILL.md` — tiny test pattern.
- `.cursor/skills/mflux-debugging/SKILL.md` — export-then-compare verification
  workflow to use for step-by-step parity checks above.
- CodeRabbit comment on `tests/model_saving/test_tiny_model_saving_fibo.py` in
  branch `tiny-tests-2` (PR review) — origin of this issue.
- Removed-for-now file: `tests/model_saving/test_tiny_model_saving_fibo.py`,
  removed in commit on `tiny-tests-2` ("Hold back FIBO tiny test pending shared
  timestep/norm parameterization") — recoverable via `git show
  42492b9:tests/model_saving/test_tiny_model_saving_fibo.py`.

**Comment by plz12345** [comment]
I think this should hold until all the issues in #577 are cleared.

**Comment by ianscrivener** [comment]
agree - FREEZE

**Comment by fxd0h** [comment]
With #577 cleared as of today, taking this one.

One note on the verification shape before I start: the heavy half here can be old-class vs new-class at width 3072 with identical weights, asserting bit-identical outputs. If the parameterized class computes the same function as the hardcoded one at the default, the PyTorch parity that already holds for main carries over unchanged. I'll run a full generation against cached FLUX.1 weights on top of that, and FIBO's if they fit on disk.
